event/request.go

```diff
@@ -0,0 +1,96 @@
+package event
+
+import (
+	"io/ioutil"
+	"mime"
+	"net/http"
+
+	"time"
+
+	"strings"
+
+	validator "gopkg.in/go-playground/validator.v9"
+)
+
+const (
+	mimeCloudEventsJSON = "application/cloudevents+json"
+)
+
+// FromRequest takes an HTTP request and returns an Event along with path
+func FromRequest(r *http.Request) (*Event, error) {
+	eventType := extractEventType(r)
+
+	mimeType, _, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+	if err != nil {
+		if err.Error() != "mime: no media type" {
+			return nil, err
+		}
+		mimeType = "application/octet-stream"
+	}
+
+	body := []byte{}
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var event *Event
+
+	if isJSONMimeType(mimeType) {
+		event, err = parseAsCloudEvent(mimeType, body)
+		if err != nil {
+			event = New(eventType, mimeType, body)
+		}
+	} else {
+		event = parseAsCloudEventBinary(New(eventType, mimeType, body), r.Header)
+	}
+
+	if eventType == TypeHTTP {
+		event.Data = NewHTTPEvent(r, event.Data)
+	}
+
+	return event, nil
+}
+
+func extractEventType(r *http.Request) Type {
+	eventType := Type(r.Header.Get("event"))
+	if eventType == "" {
+		eventType = TypeHTTP
+	}
+	return eventType
+}
+
+func parseAsCloudEventBinary(event *Event, headers http.Header) *Event {
```

Comment:
One more request here. It should not get Event as argument. It should be similar to `parseAsCloudEvent`. So the arguments should be only what's needed which is headers and `body []byte`. Event should be created inside.